File: serverless-sim/central_node/control_layer/scheduler_module/gap_solver.py
"""
GAP (Generalized Assignment Problem) Solver for Serverless Assignment
Implementation of the baseline algorithm from mobility-aware edge computing paper.
"""

import logging
import math
from typing import Dict, List, Optional, Any, Tuple
from dataclasses import dataclass

@dataclass
class GAPAssignment:
    """Result of GAP optimization"""
    user_id: str
    target_node_id: str
    node_type: str  # 'edge' or 'central'
    utility_gain: float
    estimated_latency: float
    reasoning: str

@dataclass
class GAPConfig:
    """Configuration for GAP solver"""
    # Latency parameters (ms)
    upload_size_mb: float = 300.0
    ap_to_ap_delay_per_mb: float = 0.6
    ap_to_cloud_delay_per_mb: float = 6.0
    edge_processing_rate: float = 1.0  # ms/MB
    cloud_processing_rate: float = 0.05  # ms/MB
    cold_start_penalty: float = 300.0  # ms
    
    # GAP solver options
    solver_method: str = 'greedy'  # 'greedy' or 'ilp'
    enable_memory_constraints: bool = False
    debug_logging: bool = False

class GAPSolver:
    """
    GAP-based assignment solver for edge/central nodes.
    Uses existing scheduler data structures but applies GAP optimization.
    """
    
    def __init__(self, config: GAPConfig = None):
        self.config = config or GAPConfig()
        self.logger = logging.getLogger(__name__)
        
    def calculate_latency(self, user_location: Dict[str, float], 
                         node_info: Dict[str, Any], 
                         node_type: str, 
                         is_warm_start: bool = False) -> float:
        """
        Calculate latency using simplified version of paper formula.
        Uses distance as proxy for network delay.
        """
        try:
            # Communication delay: s(u,t) * d(v_u,t, v)
            user_x, user_y = user_location.get('x', 0), user_location.get('y', 0)
            node_x, node_y = node_info.get('location', {}).get('x', 0), node_info.get('location', {}).get('y', 0)
            
            # Distance-based delay (simplified)
            distance = math.sqrt((user_x - node_x)**2 + (user_y - node_y)**2)
            
            if node_type == 'cloud':
                # Cloud: higher base delay, no cold start
                comm_delay = self.config.upload_size_mb * self.config.ap_to_cloud_delay_per_mb
                proc_delay = self.config.upload_size_mb * self.config.cloud_processing_rate
                cold_start = 0
            else:
                # Edge: distance-based delay + processing
                comm_delay = self.config.upload_size_mb * (self.config.ap_to_ap_delay_per_mb * (distance / 100.0 + 1))
                proc_delay = self.config.upload_size_mb * self.config.edge_processing_rate
                cold_start = 0 if is_warm_start else self.config.cold_start_penalty
            
            total_latency = comm_delay + proc_delay + cold_start
            return max(0, total_latency)
            
        except Exception as e:
            self.logger.error(f"Error calculating latency: {e}")
            return 1000.0  # High penalty for errors
    
    def build_profit_matrix(self, users: Dict[str, Any], 
                           edge_nodes: Dict[str, Any], 
                           central_node: Dict[str, Any]) -> Dict[str, Dict[str, float]]:
        """
        Build profit matrix H[u][v] = utility gain when assigning user u to node v.
        H[u][v] = D(u, cloud, t) - D(u, v, t)
        """
        profits = {}
        
        # Virtual cloud node for comparison
        cloud_node = {
            'location': {'x': 0, 'y': 0},
            'node_id': 'cloud'
        }
        
        for user_id, user_info in users.items():
            profits[user_id] = {}
            user_location = user_info.get('location', {'x': 0, 'y': 0})
            
            # Calculate cloud latency (baseline)
            cloud_latency = self.calculate_latency(user_location, cloud_node, 'cloud')
            
            # Calculate profit for each edge node
            for node_id, node_info in edge_nodes.items():
                edge_latency = self.calculate_latency(user_location, node_info, 'edge')
                profits[user_id][node_id] = max(0, cloud_latency - edge_latency)
            
            # Calculate profit for central node
            if central_node:
                central_latency = self.calculate_latency(user_location, central_node, 'central')
                profits[user_id]['central_node'] = max(0, cloud_latency - central_latency)
            
            # Cloud has 0 profit (baseline)
            profits[user_id]['cloud'] = 0
            
        return profits
    
    def solve_gap_greedy(self, users: Dict[str, Any], 
                        edge_nodes: Dict[str, Any], 
                        central_node: Dict[str, Any]) -> List[GAPAssignment]:
        """
        Greedy approximation for GAP.
        For each user, assign to node with maximum utility gain.
        """
        profits = self.build_profit_matrix(users, edge_nodes, central_node)
        assignments = []
        
        # All available nodes
        all_nodes = {}
        for node_id, node_info in edge_nodes.items():
            all_nodes[node_id] = {'info': node_info, 'type': 'edge'}
        if central_node:
            all_nodes['central_node'] = {'info': central_node, 'type': 'central'}
        all_nodes['cloud'] = {'info': {'location': {'x': 0, 'y': 0}}, 'type': 'cloud'}
        
        for user_id, user_info in users.items():
            best_node_id = None
            best_profit = -1
            best_type = None
            
            # Find node with maximum profit
            for node_id in all_nodes:
                profit = profits[user_id].get(node_id, 0)
                if profit > best_profit:
                    best_profit = profit
                    best_node_id = node_id
                    best_type = all_nodes[node_id]['type']
            
            if best_node_id:
                # Calculate final latency
                user_location = user_info.get('location', {'x': 0, 'y': 0})
                node_info = all_nodes[best_node_id]['info']
                estimated_latency = self.calculate_latency(user_location, node_info, best_type)
                
                assignment = GAPAssignment(
                    user_id=user_id,
                    target_node_id=best_node_id,
                    node_type=best_type,
                    utility_gain=best_profit,
                    estimated_latency=estimated_latency,
                    reasoning=f"GAP greedy: max utility gain {best_profit:.2f}ms"
                )
                assignments.append(assignment)
                
                if self.config.debug_logging:
                    self.logger.info(f"GAP assignment: {user_id} -> {best_node_id} "
                                   f"(profit: {best_profit:.2f}, latency: {estimated_latency:.2f})")
        
        return assignments
    
    def solve_gap(self, users: Dict[str, Any], 
                  edge_nodes: Dict[str, Any], 
                  central_node: Dict[str, Any] = None) -> List[GAPAssignment]:
        """
        Main GAP solver entry point.
        """
        if not users:
            return []
        
        if not edge_nodes and not central_node:
            self.logger.warning("No nodes available for GAP assignment")
            return []
        
        try:
            if self.config.solver_method == 'ilp':
                # TODO: Implement ILP solver (OR-Tools)
                self.logger.info("ILP solver not implemented, using greedy")
                return self.solve_gap_greedy(users, edge_nodes, central_node)
            else:
                return self.solve_gap_greedy(users, edge_nodes, central_node)
                
        except Exception as e:
            self.logger.error(f"GAP solver error: {e}")
            return []
    
    def get_assignment_stats(self, assignments: List[GAPAssignment]) -> Dict[str, Any]:
        """Calculate statistics for GAP assignments"""
        if not assignments:
            return {
                'total_users': 0,
                'assigned_users': 0,
                'total_utility': 0,
                'avg_latency': 0,
                'edge_assignments': 0,
                'central_assignments': 0,
                'cloud_assignments': 0
            }
        
        stats = {
            'total_users': len(assignments),
            'assigned_users': len(assignments),
            'total_utility': sum(a.utility_gain for a in assignments),
            'avg_latency': sum(a.estimated_latency for a in assignments) / len(assignments),
            'edge_assignments': sum(1 for a in assignments if a.node_type == 'edge'),
            'central_assignments': sum(1 for a in assignments if a.node_type == 'central'),
            'cloud_assignments': sum(1 for a in assignments if a.node_type == 'cloud')
        }
        
        return stats

File: simulation-ui/lib/user-management/gap-solver.js
/**
 * GAP (Generalized Assignment Problem) Solver for User Assignment
 * 
 * Implements the baseline algorithm from the paper, but uses existing
 * calculateLatency function instead of implementing new latency calculation.
 */

import { calculateLatency } from "./placement-algorithms";

/**
 * Build profit matrix H[u][v] = utility gain when assigning user u to node v
 * H[u][v] = D(u, cloud, t) - D(u, v, t) 
 * where D = latency (using existing calculateLatency function)
 */
function buildProfitMatrix(users, edgeNodes, centralNodes, latencyParams) {
  const profits = {};
  
  // Create a virtual cloud node for comparison
  const cloudNode = { id: 'cloud', x: 0, y: 0, type: 'cloud' };
  
  users.forEach(user => {
    profits[user.id] = {};
    
    // Calculate cloud latency (baseline for comparison)
    const cloudLatency = calculateLatency(user, cloudNode.id, "central", edgeNodes, [cloudNode], latencyParams);
    
    // Calculate profit for each edge node
    edgeNodes.forEach(node => {
      const nodeLatency = calculateLatency(user, node.id, "edge", edgeNodes, centralNodes, latencyParams);
      profits[user.id][node.id] = Math.max(0, cloudLatency - nodeLatency); // Utility gain
    });
    
    // Calculate profit for each central node  
    centralNodes.forEach(node => {
      const nodeLatency = calculateLatency(user, node.id, "central", edgeNodes, centralNodes, latencyParams);
      profits[user.id][node.id] = Math.max(0, cloudLatency - nodeLatency); // Utility gain
    });
    
    // Cloud has 0 profit (baseline)
    profits[user.id]['cloud'] = 0;
  });
  
  return profits;
}

/**
 * Simple greedy approximation for GAP
 * Since we don't have memory constraints in the current system,
 * this becomes a maximum weight bipartite matching problem
 */
function solveGAPGreedy(users, edgeNodes, centralNodes, latencyParams) {
  const profits = buildProfitMatrix(users, edgeNodes, centralNodes, latencyParams);
  const assignment = {};
  
  // Create list of all nodes (edge + central + cloud)
  const allNodes = [
    ...edgeNodes.map(n => ({...n, type: 'edge'})),
    ...centralNodes.map(n => ({...n, type: 'central'})),
    { id: 'cloud', type: 'cloud', x: 0, y: 0 }
  ];
  
  // For each user, find the node with maximum profit
  users.forEach(user => {
    let bestNode = null;
    let bestProfit = -1;
    
    allNodes.forEach(node => {
      const profit = profits[user.id][node.id] || 0;
      if (profit > bestProfit) {
        bestProfit = profit;
        bestNode = node;
      }
    });
    
    if (bestNode) {
      assignment[user.id] = {
        nodeId: bestNode.id,
        nodeType: bestNode.type === 'cloud' ? 'central' : bestNode.type,
        profit: bestProfit,
        latency: bestNode.type === 'cloud' 
          ? calculateLatency(user, 'cloud', "central", edgeNodes, [{ id: 'cloud', x: 0, y: 0 }], latencyParams)
          : calculateLatency(user, bestNode.id, bestNode.type, edgeNodes, centralNodes, latencyParams)
      };
    }
  });
  
  return assignment;
}

/**
 * ILP solver using simple optimization (placeholder for OR-Tools integration)
 * For now, falls back to greedy since we don't have memory constraints
 */
function solveGAPILP(users, edgeNodes, centralNodes, latencyParams) {
  // TODO: Implement actual ILP solver when memory constraints are added
  // For now, use greedy approximation
  console.log("ILP solver not implemented, using greedy approximation");
  return solveGAPGreedy(users, edgeNodes, centralNodes, latencyParams);
}

/**
 * Main GAP solver entry point
 * @param {Array} users - List of users to assign
 * @param {Array} edgeNodes - Available edge nodes
 * @param {Array} centralNodes - Available central nodes  
 * @param {Object} options - Solver options
 * @returns {Object} Assignment mapping user.id -> {nodeId, nodeType, profit, latency}
 */
export function solveGAP(users, edgeNodes, centralNodes, options = {}) {
  const {
    method = 'greedy', // 'greedy' or 'ilp'
    latencyParams = null,
    enableMemoryConstraints = false
  } = options;
  
  if (users.length === 0) {
    return {};
  }
  
  if (edgeNodes.length === 0 && centralNodes.length === 0) {
    console.warn("No nodes available for GAP assignment");
    return {};
  }
  
  try {
    switch (method) {
      case 'ilp':
        return solveGAPILP(users, edgeNodes, centralNodes, latencyParams);
      case 'greedy':
      default:
        return solveGAPGreedy(users, edgeNodes, centralNodes, latencyParams);
    }
  } catch (error) {
    console.error("GAP solver error:", error);
    // Fallback to simple nearest assignment
    const fallbackAssignment = {};
    users.forEach(user => {
      if (edgeNodes.length > 0) {
        const nearestEdge = edgeNodes[0]; // Simple fallback
        fallbackAssignment[user.id] = {
          nodeId: nearestEdge.id,
          nodeType: 'edge',
          profit: 0,
          latency: calculateLatency(user, nearestEdge.id, 'edge', edgeNodes, centralNodes, latencyParams)
        };
      }
    });
    return fallbackAssignment;
  }
}

/**
 * Utility function to get assignment statistics
 */
export function getGAPStats(assignment, users) {
  const stats = {
    totalUsers: users.length,
    assignedUsers: Object.keys(assignment).length,
    totalUtility: 0,
    avgLatency: 0,
    edgeAssignments: 0,
    centralAssignments: 0,
    cloudAssignments: 0
  };
  
  let totalLatency = 0;
  
  Object.values(assignment).forEach(assign => {
    stats.totalUtility += assign.profit || 0;
    totalLatency += assign.latency || 0;
    
    if (assign.nodeType === 'edge') stats.edgeAssignments++;
    else if (assign.nodeType === 'central') stats.centralAssignments++;
    else stats.cloudAssignments++;
  });
  
  stats.avgLatency = stats.assignedUsers > 0 ? totalLatency / stats.assignedUsers : 0;
  
  return stats;
}
